src/vsr/replica.zig

```diff
@@ -4479,112 +4413,205 @@ pub fn Replica(
+            assert(self.status == .view_change);
+            assert(self.replica_count > 1);
+            assert(self.start_view_change_quorum);
+            assert(self.do_view_change_quorum);
+            assert(self.do_view_change_from_all_replicas[self.replica] != null);
+
+            var v: ?u32 = null; // The highest `view_normal` from any replica.
+            var n: ?u64 = null; // The highest `op` for the highest `view_normal` from any replica.
+            var k: ?u64 = null; // The highest `commit_min` from any replica.
+
+            for (self.do_view_change_from_all_replicas) |received, replica| {
+                if (received) |message| {
+                    assert(message.header.command == .do_view_change);
+                    assert(message.header.cluster == self.cluster);
+                    assert(message.header.replica == replica);
+                    assert(message.header.view == self.view);
+
+                    // The view when this replica was last in normal status, which:
+                    // * may be higher than the view in any of the prepare headers.
+                    // * must be lower than the view of this view change.
+                    const view_normal = @intCast(u32, message.header.timestamp);
+                    assert(view_normal < message.header.view);
+
+                    if (replica == self.replica) {
+                        assert(view_normal == self.view_normal);
+                        assert(message.header.op == self.op);
+                        assert(message.header.commit == self.commit_min);
+                    }
+
+                    log.debug(
+                        "{}: on_do_view_change: " ++
+                            "replica={} view_normal={} op={} commit_min={}",
+                        .{
+                            self.replica,
+                            message.header.replica,
+                            view_normal,
+                            message.header.op,
+                            message.header.commit, // The `commit_min` of the replica.
+                        },
+                    );
+
+                    if (v == null or view_normal > v.?) {
+                        v = view_normal;
+                        n = message.header.op;
+                    } else if (view_normal == v.? and message.header.op > n.?) {
+                        n = message.header.op;
+                    }
+
+                    if (k == null or message.header.commit > k.?) k = message.header.commit;
+                }
+            }
+
+            assert(v.? >= self.view_normal);
+
+            // Any headers removed must be a subset of the replica's do_view_change message, and
+            // must be removed before set_op_and_commit_max() below, since that may change self.op.
+            self.discard_uncanonical_headers(v.?);
+            self.set_op_and_commit_max(n.?, k.?, "on_do_view_change");
```

Comment:
Since `self.set_op_and_commit_max` here is advancing `self.op`, after this runs `self.journal.header_for_op(self.op)` would be `null` — which temporarily violates a replica invariant.

It might be cleaner if:
- `set_op_and_commit_max` moved into `set_log_from_do_view_change_messages` immediately before or after the first `replace_header` loop
- move `discard_uncanonical_headers`  into `set_log_from_do_view_change_messages` as well, so that all mutation would happen in one place
- Then what's left of `set_op_and_commit_max_from_do_view_change_messages` function simply returns `struct { view_normal, op, commit }` (it would be renamed).